shell: add new command zyclictest

it-klinger opened the pull request:
- Name derived from Zephyr cyclictest
- Inspired by Linux realtime test program cyclictest
- Measure latency of interrupt service routine as well as thread with the help of the timer interrupt, of which the programmed point of time is known.
- Zyclictest thread is running in background while individual application under test can be running in foreground.

- The interval can be set up with optional argument -i. It should be at least double of expected worst case thread latency to get valid output. If this is not met the measurement needs to be repeated.
- The priority can also be set up. It should be the priority of the application which should run on the final product.
- The maximum size of the histogram is hard coded to 200 so far.

- After stopping measurement the maximum latency, number of errors / overflows and the histogram itself is printed.
- Histogram is printed in a form which is easy to be used by plotting programs like GNUplot or Python matplotlib.

- Example: One wants to know what is the expected maximum latency of a cooperative task with priority of -10.

  $ zyclictest start -i 400 -p -10

  $ # start your application test cases

  $ zyclictest stop

Count: 547329
                   IRQ  Thread
Max-Latency:        21      27
Errors:              0       0
Overflow:            0       0
Histogram:
...
 23                  0  547306
 24                  0       2
 25                  0       5
 26                  0       5
 27                  0       2

it-klinger committed:
shell: add new command zyclictest

- Name derived from Zephyr cyclictest
- Inspired by Linux realtime test program cyclictest
- Measure latency of interrupt service routine as well as thread with the
  help of the timer interrupt, of which the programmed point of time is
  known.
- Zyclictest thread is running in background while individual application
  under test can be running in foreground.

- The interval can be set up with optional argument -i. It should be at
  least double of expected worst case thread latency to get valid output.
  If this is not met the measurement needs to be repeated.
- The priority can also be set up. It should be the priority of the
  application which should run on the final product.
- The maximum size of the histogram is hard coded to 200 so far.

- After stopping measurement the maximum latency, number of errors /
  overflows and the histogram itself is printed.
- Histogram is printed in a form which is easy to be used by plotting
  programs like GNUplot or Python matplotlib.

- Example: One wants to know what is the expected maximum latency of a
  cooperative task with priority of -10.

  $ zyclictest start -i 400 -p -10

  $ # start your application test cases

  $ zyclictest stop

Count: 547329
                   IRQ  Thread
Max-Latency:        21      27
Errors:              0       0
Overflow:            0       0
Histogram:
...
 23                  0  547306
 24                  0       2
 25                  0       5
 26                  0       5
 27                  0       2

Signed-off-by: Andreas Klinger <ak@it-klinger.de>

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+/*
+ * zyclictest.c: Zephyr cyclictest
+ * Measure latencies of interrupt service routines and threads when woken up by a timer
+ * Collecting data in the background while application of interrest is running in the foreground
+ * Inspired by Linux cyclictest tool
```
Maybe the Kconfig help text would be a better place for this description?

it-klinger replied:
Yes and there is even more documentation meaningful which i'll add for V2.

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+
+static uint64_t zyc_cyc_sec;
+
+void zyclictest_thr(void *d0, void *d1, void *d2)
```
static

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+	zyc_data.irq_cycle = k_cycle_get_64();
+}
+
+int zyclictest_init(void)
```
static

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+	SHELL_SUBCMD_SET_END /* Array terminated. */
+);
+
+SHELL_CMD_REGISTER(zyclictest, &sub_zyclictest, "Zephyr cyclictest latency measurement", NULL);
```
do you forsee more subcommands to be creaed in addition to "start"? otherwise I guess it may make sense to just have a single `zyclictest` top level shell command

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+static int cmd_zyclictest_start(const struct shell *sh, size_t argc, char **argv)
+{
+	int ret;
+	int rv, err = 0;
```
no need to initialize `err`, also `err` `ret`? pick one

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,250 @@
+
+static struct k_timer zyc_timer;
+
+static int zyc_running;
```
should this be an atomic? makes no difference on arm but I'd imagine one could come up with a setup where it does

it-klinger replied:
good point. I'll change it.

fabiobaltieri commented:
> The idea is to be able to switch it on by demand and measure the worst case latencies with the exact circumstances of the productive application. Thus one can see if for example interrupts are executing too much code or preemption locks are taken for too long time.

Got it, I think this could use some documentation explaining this then, maybe the kconfig help text is enough but a paragraph somewhere in doc/ would not hurt.

it-klinger commented:
Hi,
is there any more feedback on the PR?
Otherwise i would prepare a V2 and submit it.

kartben commented on subsys/shell/modules/zyclictest_service.c:
fwiw the way you've formatted the code in this file is a bit unconventional for Zephyr. Especially given it's a net-new file, I would recommend applying clang-format on it. Thank you!

fabiobaltieri commented on subsys/shell/modules/zyclictest_service.c:
```diff
@@ -0,0 +1,252 @@
+	return 0;
+}
+
+int zyclictest_exit(void)
```
also static

it-klinger committed:
doc: add zyclictest shell services command

Add documentation about Zyclictest which is a tool for measuring worst case
latencies.

Signed-off-by: Andreas Klinger <ak@it-klinger.de>

it-klinger commented:
Hi,
i'm getting the following error message about RAM overflow from twister-build all the time, but i don't see how this is related to my PR.

Can anybody give me a hint, please?

This is the error:
/opt/toolchains/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/zephyr_pre0.elf section `padding_section' will not fit in region `RAM'
/opt/toolchains/zephyr-sdk-0.17.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: region `RAM' overflowed by 64528 bytes

kartben commented:
> ```
> This branch is 4 commits ahead of, 1493 commits behind zephyrproject-rtos/zephyr:main.
> ```
> 1493 is huge, rebase?

@nordicjm 
Right, but Twister CI rebases before running

kartben commented:
@nordicjm https://github.com/zephyrproject-rtos/zephyr/blob/f878a55d2e936f8b4ac3f1528009c040460e7a1d/samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay#L46-L49

Playing with the values there gives really strange behavior - for example, the below is enough to make the test pass, so I am not sure I quite understand what's with the ~64K overflow being mentioned in the error log?

```diff
diff --git a/samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay b/samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay
index aefc7bf148f..c8eb9606c48 100644
--- a/samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay
+++ b/samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay
@@ -43,9 +43,9 @@
 	};
 
 	soc {
-		sram0: memory@20006000 {
+		sram0: memory@20005000 {
 			compatible = "mmio-sram";
-			reg = <0x20006000 DT_SIZE_K(200)>;
+			reg = <0x20005000 DT_SIZE_K(200)>;
 		};
 	};
 };
```

Thanks in advance!

nordicjm commented on samples/subsys/mgmt/mcumgr/smp_svr/boards/nrf52840dk_nrf52840_ram_load.overlay:
```diff
@@ -43,9 +43,9 @@
 
 	soc {
-		sram0: memory@20006000 {
+		sram0: memory@20005000 {
```
delete this commit, it has nothing to do with this PR

nordicjm reviewed (commented):
some minor code compliance bits
